File: NetcraftServer/Plugin/NCAPI.vb
Imports System.Collections.Specialized

Namespace netcraft
    Namespace server
        Namespace api
            Public Class Logger
                Shared Sub Info(arg0 As String)
                    Form1.Log(arg0, "INFO")
                End Sub
                Shared Sub Warning(arg0 As String)
                    Form1.Log(arg0, "WARNING")
                End Sub
                Shared Sub Severe(arg0 As String)
                    Form1.Log(arg0, "ERROR")
                End Sub
            End Class
            Public Class NCSApi
                Public Shared Event BlockBreakEvent(e As events.BlockBreakEventArgs)
                Friend Shared Sub REBlockBreakEvent(e As events.BlockBreakEventArgs)
                    RaiseEvent BlockBreakEvent(e)
                End Sub
                Public Shared Event BlockPlaceEvent(e As events.BlockPlaceEventArgs)
                Friend Shared Sub REBlockPlaceEvent(e As events.BlockPlaceEventArgs)
                    RaiseEvent BlockPlaceEvent(e)
                End Sub
                Public Shared Event PlayerChatEvent(e As events.PlayerChatEventArgs)
                Friend Shared Sub REPlayerChatEvent(e As events.PlayerChatEventArgs)
                    RaiseEvent PlayerChatEvent(e)
                End Sub
                Public Shared Event PlayerMoveEvent(e As events.PlayerMoveEventArgs)
                Friend Shared Sub REPlayerMoveEvent(e As events.PlayerMoveEventArgs)
                    RaiseEvent PlayerMoveEvent(e)
                End Sub
                Public Shared Event PlayerJoinEvent(e As events.PlayerJoinEventArgs)
                Friend Shared Sub REPlayerJoinEvent(e As events.PlayerJoinEventArgs)
                    RaiseEvent PlayerJoinEvent(e)
                End Sub
                Public Shared Event PlayerLeaveEvent(e As events.PlayerLeaveEventArgs)
                Friend Shared Sub REPlayerLeaveEvent(e As events.PlayerLeaveEventArgs)
                    RaiseEvent PlayerLeaveEvent(e)
                End Sub
                Public Shared Event PlayerHealthEvent(e As events.PlayerHealthEventArgs)
                Friend Shared Sub REPlayerHealthEvent(e As events.PlayerHealthEventArgs)
                    RaiseEvent PlayerHealthEvent(e)
                End Sub
                Public Shared Event PlayerDeathEvent(e As events.PlayerDeathEventArgs)
                Friend Shared Sub REPlayerDeathEvent(e As events.PlayerDeathEventArgs)
                    RaiseEvent PlayerDeathEvent(e)
                End Sub
                Public Class Netcraft
                    Friend Shared field_a As StringCollection = New StringCollection
                    Friend Shared MainWindow As Form1
                    Private Shared field_b As List(Of Command) = New List(Of Command)
                    Friend Shared clientList As List(Of NetworkPlayer) = New List(Of NetworkPlayer)
                    Shared ReadOnly Property ConsoleCommandSender As CommandSender = New CommandSender("Server", True)
                    Shared Sub AddCommand(a As Command)
                        field_b.Add(a)
                    End Sub
                    Shared Sub RemoveCommand(a As String)
                        For Each local_a In field_b
                            If local_a.Name.ToLower = a.ToLower Then
                                field_b.Remove(local_a)
                                Exit For
                            End If
                        Next
                    End Sub
                    Shared Sub PerformCommand(arg_a As CommandSender, arg_b As Command, arg_c As String)
                        arg_b.OnCommand(arg_a, arg_b, arg_c.Split(" ").Skip(1).ToArray, arg_c)
                    End Sub
                    Shared Function GetCommands() As List(Of Command)
                        Return field_b
                    End Function
                    Shared Sub BanPlayer(a As String)
                        If field_a.Contains(a.ToLower) Then
                            Throw New Exception("Can't ban already banned player!")
                            Exit Sub
                        End If
                        For Each client In GetOnlinePlayers()
                            If client.Username.ToLower = a.ToLower Then
                                client.Kick("You have been banned from this server.")
                            End If
                        Next
                        field_a.Add(a.ToLower)
                        Form1.SaveBanlist()
                    End Sub
                    Shared Sub UnbanPlayer(a As String)
                        If Not field_a.Contains(a.ToLower) Then
                            Throw New Exception("Can't unban not banned player!")
                            Exit Sub
                        End If
                        field_a.Remove(a.ToLower)
                        Form1.SaveBanlist()
                    End Sub
                    Shared Function GetBannedPlayers() As List(Of String)
                        Dim local_a As List(Of String) = New List(Of String)
                        For Each local_b In field_a
                            local_a.Add(local_b.ToLower)
                        Next
                        Return local_a
                    End Function
                    Shared Function IsBanned(arg_a As String) As Boolean
                        Return field_a.Contains(arg_a.ToLower)
                    End Function
                    Shared Sub Broadcast(m As String)
                        'accessMainForm.Log("Broadcast: " + m)
                        'accessMainForm.Send("chat?" + m)
                        'accessMainForm(m)
                        'My.Forms.Form1.Send("chat?" + m)
                        'My.Forms.Form1.Log(m)
                        RaiseEvent dobc(m)
                    End Sub
                    Friend Shared Event dobc(m As String)
                    Shared Function GetWorld() As WorldServer
                        Return MainWindow.World
                    End Function
                    Shared Function GetPlayer(arg0 As String) As NetworkPlayer
                        For Each p In MainWindow.clientList
                            If p.Username.ToLower = arg0.ToLower Then
                                Return p
                            End If
                        Next
                        Return Nothing
                    End Function
                    Shared Function GetOnlinePlayers() As List(Of NetworkPlayer)
                        Dim s = New List(Of NetworkPlayer)
                        For Each p In MainWindow.clientList
                            s.Add(p)
                        Next
                        Return s
                    End Function
                    Shared Function GetPlayersData() As StringCollection
                        Dim s = New StringCollection
                        For Each g In IO.Directory.GetFiles(Application.StartupPath + "\playerdata\")
                            s.Add(g)
                        Next
                        Return s
                    End Function
                End Class
            End Class
        End Namespace
    End Namespace
End Namespace

File: NCore/Plugin/NCAPI.cs
using System;
using System.Collections.Generic;
using global::System.Collections.Specialized;
using System.Linq;
using System.Xml.Linq;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;
using System.Threading.Tasks;

namespace NCore.netcraft
{
    namespace server
    {
        namespace api
        {
            // Public Class Logger
            // Shared Sub Info(arg0 As String)
            // Log(arg0, "INFO")
            // End Sub
            // Shared Sub Warning(arg0 As String)
            // Log(arg0, "WARNING")
            // End Sub
            // Shared Sub Severe(arg0 As String)
            // Log(arg0, "ERROR")
            // End Sub
            // End Class
            public class PluginLogger
            {
                private Plugin plugin;

                public PluginLogger(Plugin p)
                {
                    plugin = p;
                }

                public void Info(string arg0)
                {
                    NCore.GetNCore().LogPlugin(arg0, plugin);
                }

                public void Warning(string arg0)
                {
                    NCore.GetNCore().LogPlugin(arg0, plugin, "WARNING");
                }

                public void Severe(string arg0)
                {
                    NCore.GetNCore().LogPlugin(arg0, plugin, "ERROR");
                }
            }

            public class NCSApi
            {
                public static event BlockBreakEventEventHandler BlockBreakEvent;

                public delegate void BlockBreakEventEventHandler(events.BlockBreakEventArgs e);

                internal static void REBlockBreakEvent(events.BlockBreakEventArgs e)
                {
                    BlockBreakEvent?.Invoke(e);
                }

                public static event BlockPlaceEventEventHandler BlockPlaceEvent;

                public delegate void BlockPlaceEventEventHandler(events.BlockPlaceEventArgs e);

                internal static void REBlockPlaceEvent(events.BlockPlaceEventArgs e)
                {
                    BlockPlaceEvent?.Invoke(e);
                }

                public static event PlayerChatEventEventHandler PlayerChatEvent;

                public delegate void PlayerChatEventEventHandler(events.PlayerChatEventArgs e);

                internal static void REPlayerChatEvent(events.PlayerChatEventArgs e)
                {
                    PlayerChatEvent?.Invoke(e);
                }

                public static event PlayerMoveEventEventHandler PlayerMoveEvent;

                public delegate void PlayerMoveEventEventHandler(events.PlayerMoveEventArgs e);

                internal static void REPlayerMoveEvent(events.PlayerMoveEventArgs e)
                {
                    PlayerMoveEvent?.Invoke(e);
                }

                public static event PlayerJoinEventEventHandler PlayerJoinEvent;

                public delegate void PlayerJoinEventEventHandler(events.PlayerJoinEventArgs e);

                internal static void REPlayerJoinEvent(events.PlayerJoinEventArgs e)
                {
                    PlayerJoinEvent?.Invoke(e);
                }

                public static event PlayerLeaveEventEventHandler PlayerLeaveEvent;

                public delegate void PlayerLeaveEventEventHandler(events.PlayerLeaveEventArgs e);

                internal static void REPlayerLeaveEvent(events.PlayerLeaveEventArgs e)
                {
                    PlayerLeaveEvent?.Invoke(e);
                }

                public static event PlayerHealthEventEventHandler PlayerHealthEvent;

                public delegate void PlayerHealthEventEventHandler(events.PlayerHealthEventArgs e);

                internal static void REPlayerHealthEvent(events.PlayerHealthEventArgs e)
                {
                    PlayerHealthEvent?.Invoke(e);
                }

                public static event PlayerDeathEventEventHandler PlayerDeathEvent;

                public delegate void PlayerDeathEventEventHandler(events.PlayerDeathEventArgs e);

                internal static void REPlayerDeathEvent(events.PlayerDeathEventArgs e)
                {
                    PlayerDeathEvent?.Invoke(e);
                }

                public static event BlockRightClickEventEventHandler BlockRightClickEvent;

                public delegate void BlockRightClickEventEventHandler(events.BlockRightClickEvent e);

                internal static void REBlockRightClickEvent(events.BlockRightClickEvent e)
                {
                    BlockRightClickEvent?.Invoke(e);
                }

                public static event PlayerLoginEventEventHandler PlayerLoginEvent;

                public delegate void PlayerLoginEventEventHandler(events.PlayerLoginEventArgs e);

                internal static void REPlayerLoginEvent(events.PlayerLoginEventArgs e)
                {
                    PlayerLoginEvent?.Invoke(e);
                }
            }

            public class Netcraft
            {
                internal static StringCollection field_a = new StringCollection();
                private static List<Command> field_b = new List<Command>();
                internal static List<NetworkPlayer> clientList = new List<NetworkPlayer>();

                public static CommandSender ConsoleCommandSender { get; private set; } = new CommandSender("Server", true);

                public static void AddCommand(Command a)
                {
                    field_b.Add(a);
                }

                public static void RemoveCommand(string a)
                {
                    foreach (var local_a in field_b)
                    {
                        if ((local_a.Name.ToLower() ?? "") == (a.ToLower() ?? ""))
                        {
                            field_b.Remove(local_a);
                            break;
                        }
                    }
                }

                public static void PerformCommand(CommandSender arg_a, Command arg_b, string arg_c)
                {
                    arg_b.OnCommand(arg_a, arg_b, arg_c.Split(" ").Skip(1).ToArray(), arg_c);
                }

                public static List<Command> GetCommands()
                {
                    return field_b;
                }

                public static async Task BanPlayer(string a)
                {
                    if (field_a.Contains(a.ToLower()))
                    {
                        throw new Exception("Can't ban already banned player!");
                    }

                    foreach (var client in GetOnlinePlayers())
                    {
                        if ((client.Username.ToLower() ?? "") == (a.ToLower() ?? ""))
                        {
                            await client.Kick(NCore.GetNCore().lang.get("error.banned"));
                        }
                    }

                    field_a.Add(a.ToLower());
                    NCore.GetNCore().SaveBanlist();
                }

                public static async Task UnbanPlayer(string a)
                {
                    if (!field_a.Contains(a.ToLower()))
                    {
                        throw new Exception("Can't unban not banned player!");
                        return;
                    }

                    field_a.Remove(a.ToLower());
                    NCore.GetNCore().SaveBanlist();
                }

                public static List<string> GetBannedPlayers()
                {
                    var local_a = new List<string>();
                    foreach (string local_b in field_a)
                        local_a.Add(local_b.ToLower());
                    return local_a;
                }

                public static bool IsBanned(string arg_a)
                {
                    return field_a.Contains(arg_a.ToLower());
                }

                public static async Task Broadcast(string m)
                {
                    // accessMainForm.Log("Broadcast: " + m)
                    // accessMainForm.Send("chat?" + m)
                    // accessMainForm(m)
                    // My.Forms.Form1.Send("chat?" + m)
                    // My.Forms.Form1.Log(m)
                    dobc?.Invoke(m);
                }

                internal static event dobcEventHandler dobc;

                internal delegate void dobcEventHandler(string m);

                public static WorldServer GetWorld()
                {
                    return NCore.GetNCore().World;
                }

                public static NetworkPlayer GetPlayer(string arg0)
                {
                    foreach (var p in clientList)
                    {
                        if ((p.Username.ToLower() ?? "") == (arg0.ToLower() ?? ""))
                        {
                            return p;
                        }
                    }

                    return null;
                }

                public static List<NetworkPlayer> GetOnlinePlayers()
                {
                    var s = new List<NetworkPlayer>();
                    foreach (var p in clientList)
                        s.Add(p);
                    return s;
                }

                public static StringCollection GetPlayersData()
                {
                    var s = new StringCollection();
                    foreach (var g in System.IO.Directory.GetFiles("./playerdata/"))
                        s.Add(g);
                    return s;
                }
            }

            public class Config
            {
                internal string text;

                public static Config Read(string a)
                {
                    var cfg = new Config();
                    cfg.text = System.IO.File.ReadAllText(a, System.Text.Encoding.UTF8);
                    cfg.path = a;
                    return cfg;
                }

                internal string path;

                public void Save()
                {
                    System.IO.File.WriteAllText(path, text, System.Text.Encoding.UTF8);
                }

                public string Parse(string variable)
                {
                    return global::NCore.Config.GetValue(variable, text.Replace(Constants.vbCrLf, Constants.vbLf), "");
                }
            }
        }
    }
}